### PR TITLE
FIX-#3276: fix ZeroDivision error during data import operation call

### LINKED ISSUE #3276: Data import operations fail if empty frame should be read
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
- **Modin version**: d03ed30868fb042bbc7282b9c9574145e1228eeb
- **Python version**: 3.8.10
- **Code we can use to reproduce**:
```
import modin.pandas as pd
from io import StringIO
import os

test_filename = "test.csv"
data = """col1,col2
1,2
3,4
"""

try :
    with open(test_filename, "w") as f:
        f.write(data)
    
    pd.read_csv(test_filename, usecols=["col1"], index_col="col1")
finally:
    os.remove(test_filename)
```

### Describe the problem

`read_csv` fails during metadata calculation.

### Source code / logs

```
Traceback (most recent call last):
  File "zero_div.py", line 15, in <module>
    pd.read_csv(test_filename, usecols=["col1"], index_col="col1")
  File "/modin/modin/pandas/io.py", line 131, in read_csv
    return _read(**kwargs)
  File "/modin/modin/pandas/io.py", line 59, in _read
    pd_obj = FactoryDispatcher.read_csv(**kwargs)
  File "/modin/modin/data_management/factories/dispatcher.py", line 172, in read_csv
    return cls.__factory._read_csv(**kwargs)
  File "/modin/modin/data_management/factories/factories.py", line 206, in _read_csv
    return cls.io_cls.read_csv(**kwargs)
  File "/modin/modin/engines/base/io/file_dispatcher.py", line 67, in read
    query_compiler = cls._read(*args, **kwargs)
  File "/modin/modin/engines/base/io/text/csv_dispatcher.py", line 112, in _read
    column_widths, num_splits = cls._define_metadata(
  File "/modin/modin/engines/base/io/text/text_file_dispatcher.py", line 422, in _define_metadata
    column_chunksize = compute_chunksize(df, num_splits, axis=1)
  File "/modin/modin/data_management/utils.py", line 70, in compute_chunksize
    col_chunksize = get_default_chunksize(len(df.columns), num_splits)
  File "/modin/modin/data_management/utils.py", line 37, in get_default_chunksize
    length // num_splits if length % num_splits == 0 else length // num_splits + 1
ZeroDivisionError: integer division or modulo by zero
```

### PR DESCRIPTION
**Pull request by amyskov**
Signed-off-by: Alexander Myskov <alexander.myskov@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3276   
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 1]
```diff
@@ -107,7 +107,8 @@ def _read(cls, filepath_or_buffer, **kwargs):
         # Max number of partitions available
         num_partitions = NPartitions.get()
         # This is the number of splits for the columns
-        num_splits = min(len(column_names), num_partitions)
+        cols_number = len(column_names)
+        num_splits = min(cols_number if cols_number else 1, num_partitions)
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_dispatcher.py:111]:
```suggestion
        num_splits = min(len(column_names) or 1, num_partitions)
```

Comment (anmyachev) [reply modin/engines/base/io/text/csv_dispatcher.py:111]: Why is the minimum number of columns - 1?

Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py]: This is done to avoid division by zero [here](https://github.com/modin-project/modin/blob/master/modin/data_management/utils.py#L37) (for full traceback see issue header).

File: modin/engines/base/io/text/csv_dispatcher.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def _read(cls, filepath_or_buffer, **kwargs):
         # Max number of partitions available
         num_partitions = NPartitions.get()
         # This is the number of splits for the columns
-        num_splits = min(len(column_names), num_partitions)
+        num_splits = min(len(column_names) or 1, num_partitions)
```
Comment (YarShev) [inline modin/engines/base/io/text/csv_dispatcher.py:110]: Is this adding really required?

File: modin/backends/pandas/parsers.py [hunk 3]
```diff
@@ -153,8 +153,11 @@ def get_dtypes(cls, dtypes_ids):
             pandas.Series where index is columns names and values are
             columns dtypes.
         """
+        partitions_dtypes = cls.materialize(dtypes_ids)
+        if all([len(dtype) == 0 for dtype in partitions_dtypes]):
```
Comment (YarShev) [inline modin/backends/pandas/parsers.py:157]: Can you elaborate why this change fixes the issue?

File: modin/pandas/test/test_io.py [hunk 4]
```diff
@@ -1115,6 +1115,22 @@ def get_internal_df(df):
         for name1, name2 in zip(get_internal_df(read_df).index.names, [None, "a"]):
             assert name1 == name2
 
+    def test_read_csv_empty_frame(
+        self,
+        request,
+    ):
+        if request.config.getoption("--simulate-cloud").lower() != "off":
```
Comment (YarShev) [inline modin/pandas/test/test_io.py:1122]: Can you consider this comment?
https://github.com/modin-project/modin/issues/3295#issuecomment-889047542

File: modin/backends/pandas/parsers.py (hunk 3 above)
Comment (amyskov) [reply modin/backends/pandas/parsers.py:157]: Yes, this change is needed because when empty frame should be read, each worker will read empty frame and it's `dtypes` will be empty `Series` that corresponds to `dtypes=None`. If we skip this check, `dtypes` will values will be calculated incorrectly because `partitions_dtypes` contains no columns.

File: modin/engines/base/io/text/csv_dispatcher.py (hunk 2 above)
Comment (amyskov) [reply modin/engines/base/io/text/csv_dispatcher.py:110]: Yes, we need to set non-zero `num_splits` value to avoid `ZeroDivision` error [here](https://github.com/modin-project/modin/blob/master/modin/data_management/utils.py#L37).

File: modin/pandas/test/test_io.py (hunk 4 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Yes, fixed.

**Review by YarShev — DISMISSED** [review]
@amyskov , LGTM assuming tests pass. @anmyachev , any other comments?

**Comment by anmyachev** [comment]
@amyskov you are changing the way define partitions for `read_csv` function only, but this problem is more general. It is necessary to generalize changes at least to all I/O functions.

**Comment by YarShev** [comment]
@anmyachev , good notice! We have some places where num_splits is computed in depend on the length of columns. There might be an issue when reading empty frames from other files. @amyskov , can you check it?

**Comment by amyskov** [comment]
> @anmyachev , good notice! We have some places where num_splits is computed in depend on the length of columns. There might be an issue when reading empty frames from other files. @amyskov , can you check it?

Yes, `num_splits` check was moved to the higher level function, so now it affects all IO functions.

File: modin/data_management/utils.py [hunk 5]
```diff
@@ -59,6 +59,8 @@ def compute_chunksize(df, num_splits, default_block_size=32, axis=None):
         If axis is 1 or 0, returns an integer number of rows/columns to split the
         DataFrame. If axis is None, returns a tuple containing both.
     """
+    # `num_splits` should be a non-zero value
+    num_splits = num_splits if num_splits > 0 else 1
```
Comment (anmyachev) [inline modin/data_management/utils.py:63]: `num_splits` is also used in functions:
* `_define_metadata`;
* `split_result_of_axis_func_pandas`;
* Based on this variable, a `column_widths` variable for `build_partition` function is determined (For example, in `JSONDispatcher._read`)

Should we also change these places for consistency?

It would be nice to add examples to some other I/O functions (For example, `read_json`).

Comment (amyskov) [reply modin/data_management/utils.py]: In the case of empty frame reading, `num_splits` will be redefined to 1 in the `_define_metadata` and then processed normally further, but i think it will be more correct to set correct `num_splits` value in the moment of `num_splits` definition, so previous approach was used.
I suggest to fix in this PR only `read_csv` function as it mentioned in the issue. Actually, what you mean under "examples"?

Comment (anmyachev) [reply modin/data_management/utils.py]: Oh, I wanted to say `add more IO tests` :)

Comment (anmyachev) [reply modin/data_management/utils.py]: There are a few more places where similar lines can be found:
* https://github.com/modin-project/modin/blob/54898ef78e83209fba3233606841ca988e531aee/modin/engines/base/io/text/excel_dispatcher.py#L163
* https://github.com/modin-project/modin/blob/54898ef78e83209fba3233606841ca988e531aee/modin/engines/base/io/text/fwf_dispatcher.py#L157

I believe that the same error exists for them. Why make changes for them separately if the problematic mechanism (`num_splits` definition) is the same?

Comment (amyskov) [reply modin/data_management/utils.py]: Yes, you are right, the same issue can occur for another dispatchers, but for excel dispatcher we wouldn't be capable to verify this fix because of https://github.com/modin-project/modin/issues/3305, so additional PR should be done anyway. But, i think, it should be fine to add these fixes anyway, so they were added.

File: modin/engines/base/io/text/csv_glob_dispatcher.py [hunk 6]
```diff
@@ -184,7 +184,7 @@ def _read(cls, filepath_or_buffer, **kwargs):
             # Max number of partitions available
             num_partitions = NPartitions.get()
             # This is the number of splits for the columns
-            num_splits = min(len(column_names), num_partitions)
+            num_splits = min(len(column_names) or 1, num_partitions)
```
Comment (anmyachev) [inline modin/engines/base/io/text/csv_glob_dispatcher.py:187]: Is it possible to move this code into a separate function so that this code can be reused?

Comment (amyskov) [reply modin/engines/base/io/text/csv_glob_dispatcher.py]: Yes, this and other related code was moved to the `_define_metadata` function.

File: modin/engines/base/io/text/json_dispatcher.py [hunk 7]
```diff
@@ -119,3 +108,44 @@ def _read(cls, path_or_buf, **kwargs):
         )
         new_frame.synchronize_labels(axis=0)
         return cls.query_compiler_cls(new_frame)
+
+    @classmethod
+    def _define_metadata(
```
Comment (anmyachev) [inline modin/engines/base/io/text/json_dispatcher.py:113]: We already have `_define_metadata` function in `text_file_dispatcher.py`.

Why create another function with the same name in `json_dispatcher.py`?
Shouldn't they be doing the same thing?

Comment (anmyachev) [reply modin/engines/base/io/text/json_dispatcher.py:113]: It seems to me that there is a bug in `_define_metadata` function in `json_dispatcher.py`.

Just try with next parameters: `num_splits == 4, len(column_names) == 80 and column_chunksize == 32`.

It turns out that `column_widths=[32, 32, 32, -16]`, but should be equals to `[32, 32, 16, 0]`

Comment (anmyachev) [reply modin/engines/base/io/text/json_dispatcher.py:113]: So we can use `_define_metadata` from `text_file_dispatcher.py` in all place.

I suggest first fixing the found bug in a separate pull request, and then merge this one.

**Comment by anmyachev** [comment]
@amyskov please don't forget to update issue, since you are fixing a more global problem.

**Comment by YarShev** [comment]
@amyskov , please, ping me when the PR is ready for review.

File: modin/engines/base/io/text/json_dispatcher.py (hunk 7 above)
Comment (amyskov) [reply modin/engines/base/io/text/json_dispatcher.py:113]: You are right, it's a bug, corresponding issue was created https://github.com/modin-project/modin/issues/3313.

**Comment by amyskov** [comment]
@YarShev , PR is ready for review.

**Comment by YarShev** [comment]
@amyskov , please, consider the discussion from #3314.

**Review by YarShev — APPROVED** [review]
@amyskov , LGTM, thanks!
